pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,320 @@
+            source_kernels=(base_kernel,), densities=(1,))
+    for int_g in int_gs:
+        for source_kernel in int_g.source_kernels:
+            if source_kernel.get_base_kernel() != base_kernel:
+                return False
+        if common_int_g != int_g.copy(target_kernel=base_kernel,
+                source_kernels=(base_kernel,), densities=(1,)):
+            return False
+    return True
+
+
+def _convert_kernel_to_poly(kernel, axis_vars):
+    """Converts a :class:`sumpy.kernel.Kernel` object to a polynomial.
+    A :class:`sumpy.kernel.Kernel` represents a derivative operator
+    and the derivative operator is converted to a polynomial with
+    variables given by `axis_vars`.
+
+    For eg: for target y and source x the derivative operator,
+    d/dx_1 dy_2 + d/dy_1 is converted to -y_2 * y_1 + y_1.
+    """
+    if isinstance(kernel, AxisTargetDerivative):
+        poly = _convert_kernel_to_poly(kernel.inner_kernel, axis_vars)
+        return axis_vars[kernel.axis]*poly
+    elif isinstance(kernel, AxisSourceDerivative):
+        poly = _convert_kernel_to_poly(kernel.inner_kernel, axis_vars)
+        return -axis_vars[kernel.axis]*poly
+    return 1
+
+
+def _convert_source_poly_to_int_g(poly, orig_int_g, axis_vars):
+    """This does the opposite of :func:`_convert_kernel_to_poly`
+    and converts a polynomial back to a source derivative
+    operator. First it is converted to a :class:`sumpy.kernel.Kernel`
+    and then to a :class:`~pytential.symbolic.primitives.IntG`.
+    """
+    to_pymbolic = SympyToPymbolicMapper()
+
+    orig_kernel = orig_int_g.source_kernels[0]
+    source_kernels = []
+    densities = []
+    for monom, coeff in poly.terms():
+        kernel = orig_kernel
+        for idim, rep in enumerate(monom):
+            for _ in range(rep):
+                kernel = AxisSourceDerivative(idim, kernel)
+        source_kernels.append(kernel)
+        densities.append(to_pymbolic(coeff) * (-1)**sum(monom))
+    return orig_int_g.copy(source_kernels=tuple(source_kernels),
+            densities=tuple(densities))
+
+
+def _convert_target_poly_to_int_g(poly, orig_int_g, rhs_int_g):
+    """This does the opposite of :func:`_convert_kernel_to_poly`
+    and converts a polynomial back to a target derivative
+    operator. It is applied to a :class:`~pytential.symbolic.primitives.IntG`
+    object and returns a new instance.
+    """
+    to_pymbolic = SympyToPymbolicMapper()
+
+    result = 0
+    for monom, coeff in poly.terms():
+        kernel = orig_int_g.target_kernel
+        for idim, rep in enumerate(monom):
+            for _ in range(rep):
+                kernel = AxisTargetDerivative(idim, kernel)
+        result += orig_int_g.copy(target_kernel=kernel,
+                source_kernels=rhs_int_g.source_kernels,
+                densities=rhs_int_g.densities) * to_pymbolic(coeff)
+
+    return result
+
+
+class ConvertDensityToSourceExprCoeffMap(Mapper):
```

Comment:
- What does this thing do?
- Is it meaningfully different from pymbolic's built-in `CoefficientFinder`? https://github.com/inducer/pymbolic/blob/main/pymbolic/mapper/coefficient.py

Reply:
Done in 71d15401611400bf47c5d27d99dff394e3dee0b3